Simplified bytes allocation

nazar-pc opened the pull request:
Implements #5449

In one of the projects I'm working on it allows to reduce following:
```livescript
msgArrPtr     = supercop._malloc(message.length)
msgArr        = new Uint8Array(supercop.HEAPU8.buffer, msgArrPtr, message.length)
pubKeyArrPtr  = supercop._malloc(32)
pubKeyArr     = new Uint8Array(supercop.HEAPU8.buffer, pubKeyArrPtr, 32)
privKeyArrPtr = supercop._malloc(64)
privKeyArr    = new Uint8Array(supercop.HEAPU8.buffer, privKeyArrPtr, 64)
sigPtr        = supercop._malloc(64)
sig           = new Uint8Array(supercop.HEAPU8.buffer, sigPtr, 64)
msgArr.set(message)
pubKeyArr.set(publicKey)
privKeyArr.set(privateKey)
supercop
    .._ed25519_sign(sigPtr, msgArrPtr, message.length, pubKeyArrPtr, privKeyArrPtr)
    .._free(msgArrPtr)
    .._free(pubKeyArrPtr)
    .._free(privKeyArrPtr)
    .._free(sigPtr)
new Uint8Array(sig)
```

To this:
```livescript
message    = supercop.allocateBytes(0, message)
publicKey  = supercop.allocateBytes(0, publicKey)
privateKey = supercop.allocateBytes(0, privateKey)
signature  = supercop.allocateBytes(64)
supercop
  .._ed25519_sign(signature, message, message.length, publicKey, privateKey)
  ..freeBytes()
signature.get()
```

Code is written in LiveScript, but it shouldn't be too difficult to read (compiled JavaScript version of the first snippet is even longer).

Besides obvious readability and usability improvements it also allows to reduce the amount of code necessary to write during module consumption, which results in smaller browser bundles.

Let's stick to implementation discussion here and use #5449 to discuss the concept in general.

If this looks good, I'll update documentation and will probably even add some tests if someone helps me with that. Also I think this can be potentially hidden under compile-time option to reduce a few bytes for those who don't need this wrapper functions.

juj commented:
I don't think we'd want to do this in here. You can create a JS library with `--js-library` that does this or a `--pre-js` to add this functionality, or just add this code in your own Module object, so there should be at least three clean ways to add these types of helpers. Agreed that it looks much simpler for users to code that way, but I think we'll want to optimize towards minimum compiled code size from the compiler, since not all users will need to use these.

nazar-pc commented:
What about optional compile-time flag for this feature?
I believe this would be useful for many projects and having canonical efficient implementation might benefit those who actually need it.
No doubt, it is possible to implement this without touching Emscripten core, but copy-pasting the same 50 lines in every project is tedious.

kripken commented:
Perhaps it makes sense to add a new API for this. It should be optional though in that it should be removeable by the compiler if it's not used (so in particular it shouldn't be exported by default on Module, I think). Or perhaps it could just be a separate js file (wrappingUtils.js?) that would need to be included if the user wants it?

For concrete API discussion, the allocation side might already be doable with the `allocate` function (which can be told to use malloc, and given a buffer of data to write). So maybe all we need is to add a method that gathers allocations and frees them all at once?

nazar-pc commented:
> Or perhaps it could just be a separate js file (wrappingUtils.js?) that would need to be included if the user wants it?

Separate file with compile-time option (off by default) sounds good to me

As for `allocate` function, to be honest, it is not very clear from documentation what it does exactly, how, and when should be preferred.

kripken commented:
Yes, `allocate` is mostly for internal use, it's not a very friendly API.

Overall I'm not sure if we should add this to the code, or maybe we could put this as sample code on the docs. Both seem reasonable.

nazar-pc commented:
I've made a series of improvements to initial implementation:
* added `Module.createPointer()` - is used internally, but might be generally useful
* `allocateBytes()` can be initialized (and later set) with any TypedArray (only Uint9Array was supported initially) or string
* added `allocatePointer()` function
* added `dereference()` method on pointer

This allows us to make things extremely easy and explicit:
```c
void func (int **i) {...}
```

```javascript
var pointer_to_pointer = Module.allocatePointer();
Module._func(pointer_to_pointer);
// By default dereferences pointers (4 bytes), but custom size can be specified if needed
var pointer = pointer_to_pointer.dereference(1);
console.log(pointer.get());
var another_pointer_to_pointer = Module.allocatePointer(pointer);
```

I think users will appreciate how much it is possible to do with how little effort. I honestly believe this should be somewhere in Emscripten core, even if behind a flag and turned off by default.

kripken commented:
Yeah, this is interesting. Maybe it would be a good idea to open a discussion on the mailing list to see how people feel. For example if people start using your APIs then it would motivate including them.

But another factor here to consider is that wasm is considering adding convenience APIs for wrapping to JS. This would overlap with what you're suggesting here, which suggests that maybe we should wait for that to appear before adding anything to core emscripten.

nazar-pc commented:
I've created 3 more crypto-related libraries using Emscripten since initial PR submission and was using these exact APIs in each of them (also made some improvements and optimizations along the way).

So I've just rebased this PR and turned it into new option `-s SIMPLE_ALLOCATION_API=1`. This way it will not introduce overhead for projects that do not use it and will be there for those who need it.

Here is one project that uses these APIs heavily (note that `Module.allocateBytes()` is aliased as `allocate()`): https://github.com/nazar-pc/noise-c.wasm/blob/master/src/index.ls

If it looks good, I can add some tests (in test_other.py?) and update docs.

kripken commented:
I like this. And I think it's good to keep it as an option so it doesn't add any code by default.

But I would like to see more evidence other people would use it before merging, and feedback on the API design. Perhaps write up a few tests that show the usage, and we can discuss here or on the mailing list?
